### PR TITLE
DSA: Add FP8/MXFP8 and compressed Top-K indexer paths

### PR DESCRIPTION
**Pull request by jiayus-nvidia**
## Summary

This PR adds FP8/MXFP8 support to the DSA indexer score paths and introduces an
SM100 compressed indexer-forward + Top-K pipeline.

The SM100 indexer score implementation is shared by dense forward and dense
recompute, while the compressed path produces Top-K indices, selected logits,
optional fused softmax, and optional LSE without materializing the full dense
score tensor.

SM100 MXFP8 indexer paths support `qhead_per_kv_head ∈ {32, 64}`. The QH32
epilogue is compile-time specialized separately so the existing QH64 hot path
and launch topology remain unchanged.

### FP8/MXFP8 score paths

- Add an SM90 FP8 indexer path using E4M3 Q/K with per-token/head FP32
  descales.
- Add SM100 MXFP8 indexer and dense-attention score recompute paths using E4M3
  Q/K with packed E8M0 block scales.
- Share unified SM100 indexer score kernels between dense forward and dense
  indexer score recompute.
- Support BSHD and THD inputs, including compact padded MXFP8 scale layouts for
  variable-length inputs.
- Preserve the existing BF16 paths.

### Compressed indexer forward

- Add the SM100-only `indexer_forward_top_k_wrapper`.
- Return INT32 Top-K indices, FP32 selected logits, fused FP32 softmax by
  default, and optional FP32 LSE.
- Support BF16 and MXFP8 inputs, BSHD and THD layouts, and local or global
  output indices; global KV indices are returned by default.
- With `deterministic=True`, resolve exact-value ties at the K-th boundary
  toward the smallest local KV indices. This makes the selected set
  reproducible but does not sort the output slots.
- Support caller-owned candidate, candidate-offset, index, logit, softmax, and
  LSE buffers.
- Support optional BF16 BSHD microbatching and dedicated BSHD/THD
  candidate-buffer sizing helpers.
- Fix large THD candidate-buffer indexing.

### Runtime and backward integration

- Keep allocations, conversions, copies, launches, and optional local-to-global
  conversion on one caller-selected CUDA stream. The launch path does not add
  cross-stream synchronization; callers own input readiness and tensor
  lifetime.
- Validate THD device metadata structurally without copying device-resident
  prefixes to the host. Callers guarantee valid sequence, scale, causal-offset,
  and candidate-prefix values, alignment, and backing storage.
- Allow the fused forward softmax to be passed directly as sparse backward
  `index_score`, avoiding indexer-score recompute and a separate softmax.
- Support the packed BF16 THD-to-backward path through zero-copy `B=1` BSHD
  views with global Top-K indices.
- Sparse backward consumes and overwrites `attn_score` and `index_score`;
  callers must copy either tensor if it must be preserved.
- Remove the superseded SM100 decode KV-split and partial-LSE merge path.
- Update documentation, reference helpers, and DSA coverage.

## Validation

Validated on an NVIDIA B200 with CUDA 13.2, PyTorch 2.11, and cuDNN backend
9.20:

```bash
cd test/python

PYTHONPATH=../../python pytest -q \
  fe_api/dsa/test_DSA_mxfp8_scale_utils.py \
  fe_api/dsa/test_DSA_indexer_forward.py \
  fe_api/dsa/test_DSA_dense_score_recompute.py
# 28 passed, 6 skipped

PYTHONPATH=../../python pytest -q \
  fe_api/dsa/test_DSA_mxfp8_scale_utils.py \
  fe_api/dsa/test_DSA_indexer_forward.py \
  fe_api/dsa/test_DSA_dense_score_recompute.py \
  -k mxfp8
# 22 passed, 12 deselected
```

The full DSA suite completed with `70 passed, 7 skipped, 4 failed`. The four
failures reproduce identically on current `upstream/develop` in an independent
worktree:

- two causal standalone Top-K cases with `top_k=705`;
- two SM100 sparse-backward D576 cases.

Formatting and static checks:

```bash
git diff --check upstream/develop...HEAD
git diff --name-only upstream/develop...HEAD -- '*.py' \
  | xargs python3 -m black --check -l 160
git diff --name-only upstream/develop...HEAD -- '*.py' \
  | xargs python3 -m py_compile
```

QH64 B200 performance was remeasured with the same inputs, 10 warmups, and 100
CUDA-event iterations:

| Path | Before | After |
|---|---:|---:|
| Dense indexer | 0.123456 ms | 0.123456 ms |
| Compressed stage 1 | 0.109088 ms | 0.109088 ms |
| Combined Top-K + softmax | 0.227872 ms | 0.227872 ms |

## Limitations / notes

- The combined compressed forward + Top-K path is SM100-only. SM100 dense and
  compressed indexer paths currently require MQA (`H_kv = 1`).
- Indexer kernels require `head_dim = 128`. SM100 MXFP8 additionally requires
  `qhead_per_kv_head ∈ {32, 64}` and `sf_vec_size = 32`.
- Explicit microbatching is limited to BF16 BSHD and cannot be combined with
  MXFP8, LSE, or explicit `q_causal_offsets`.
- Top-K slot order is not guaranteed to be sorted. Padded slots use index `-1`,
  logit `-inf`, and softmax `0`.
- Standalone and combined indexer Top-K both require
  `1 <= top_k <= 2048`, although they use different implementations.
- FP8/MXFP8 indexer backward is not included; current backward Q/K/W gradient
  kernels require BF16 operands.
- BSHD with explicit `q_causal_offsets` computes candidate offsets eagerly and
  is not CUDA-graph-capturable. THD graph capture requires explicit maximum
  sequence lengths, caller-provided candidate offsets, and preallocated
  outputs. Local indices are required only for strict zero-transient-allocation
  capture; global conversion is capturable but records temporaries in the graph
  pool.

## Summary by CodeRabbit

* **New Features**
  * Added compressed indexer-forward with integrated Top‑K selection (including fused softmax), optional LSE output, global index support, and deterministic tie-breaking.
  * Expanded precision support for dense workflows: FP8 on newer GPUs and MXFP8 on SM100+ with packed scale handling.
  * Added new buffer-sizing helpers and enhanced APIs for both standard and variable-length (THD) inputs.
* **Documentation**
  * Updated DSA docs/README with revised GPU/precision constraints and output semantics (including LSE and masking behavior).
* **Bug Fixes**
  * Improved consistency of denom placeholder allocation behavior for SM100+ compressed paths.
* **Tests**
  * Added/expanded regression coverage for compressed Top‑K, FP8/MXFP8, THD scaling/packing, LSE correctness, masking, and deterministic mode.

### CODE REVIEW

**Comment by Anerudhan** [comment]
@cudnn-ci-bot run

**Comment by jiayus-nvidia** [comment]
WIP for adding deterministic and e2e test

File: python/cudnn/deepseek_sparse_attention/score_recompute/_interface_sm100.py [hunk 1]
```diff
@@ -942,29 +1246,68 @@ def _dense_attn_score_recompute(
         denom_shape = (q.shape[0],) if is_varlen else (bs, seqlen_q)
         with _torch_stream_context(current_stream):
             denom_out = torch.empty(denom_shape, dtype=torch.float32, device=device)
+    if precision == "mxfp8" and ((head_dim == 512 and qhead_per_kv_head >= 128) or (head_dim == 128 and qhead_per_kv_head == 64)):
+        with _torch_stream_context(current_stream):
+            out.zero_()
+            denom_out.zero_()
```
[the remark this answers is not in the document]
Comment (jiayus-nvidia) [reply python/cudnn/deepseek_sparse_attention/score_recompute/_interface_sm100.py:1255]: Thanks for flagging this. I traced the D512/QH64 path, and this configuration does not accumulate into the same `mDenom` element.

For D512/QH64, `use_dual_attention_epilogue` assigns the two epilogue warpgroups to different query-token stages. `_epilogue_attention_dense_single_q(0/1)` computes distinct `q_token_idx` values and finishes with a direct overwrite:

```python
mDenom[q_token_idx] = global_sum
```

The `atomic_add` paths are the head-split epilogues: D128/QH64 (`_epilogue_attention_dense_q_head_half`) and D512/QH128 (`_epilogue_attention_dense_h128_head_half`). Those are exactly the configurations covered by the existing `denom_out.zero_()` condition.

I also reran the targeted D512/QH64 MXFP8 reference case:

```
PYTHONPATH=../../python pytest -q fe_api/dsa/test_DSA_dense_score_recompute.py -k "thd_mxfp8_compact_padded_scales and attention_d512"
# 1 passed
```

So expanding the denom zero-initialization condition to D512/QH64 would add an unnecessary memset and is not needed for correctness. The early `out.zero_()` for the existing atomic configurations is redundant with the pre-launch zero and can be cleaned up independently, but it does not imply a D512/QH64 denom initialization issue.

File: python/cudnn/deepseek_sparse_attention/indexer_forward/_compressed_top_k_sm100.py [hunk 2]
```diff
@@ -0,0 +1,1522 @@
...
+        alloc_device = torch.device("cuda", device_index)
+    else:
+        device_index = device.index
+        alloc_device = device
+    key = (alloc_device.type, device_index, shape)
+    cached = _denom_placeholder_cache.get(key)
+    if cached is None or cached.device != alloc_device:
+        cached = torch.empty(shape, dtype=torch.float32, device=alloc_device)
+        _denom_placeholder_cache[key] = cached
+    return cached
```
[the remark this answers is not in the document]
Comment (jiayus-nvidia) [reply python/cudnn/deepseek_sparse_attention/indexer_forward/_compressed_top_k_sm100.py]: Fixed in `734dbb9f`. Both the compressed and dense placeholder caches now use stable power-of-two capacity buckets keyed by device, rank, and capacity, then return a slice/view for the requested shape. Buckets are retained instead of replaced so CUDA Graph captures keep stable addresses, and first creation is lock-protected. Added coverage for within-bucket reuse, capacity growth, and rank isolation; the targeted SM100 tests pass.

**Review by Anerudhan — APPROVED** [review]

**Comment by Anerudhan** [comment]
@cudnn-ci-bot run
